[CBRD-26256] Support adding FK constraints on table creation

youngjun9072 opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-26256>](http://jira.cubrid.org/browse/CBRD-26256)

### Purpose

HA 모드에서 외래키 사용시 다음 제약사항에 대한 CREATE TABLE, ALTER TABLE의 변경입니다.
* 참조 테이블(복제 ON) <- 외래키 테이블(복제 ON) OK
* 참조 테이블(복제 ON) <- 외래키 테이블(복제 OFF) OK
* 참조 테이블(복제 OFF) <- 외래키 테이블(복제 OFF) OK
* 참조 테이블(복제 OFF) <- 외래키 테이블(복제 ON) ERROR

### Implementation

CREATE TABLE 시 호출되는 테이블에서는 복제 옵션이 ON이라면 다음 제약사항을 확인하며, 
특히 3번 단계는 이번 이슈에서 추가된 기능입니다.
1. HA 모드 여부
2. 복제 키 존재 여부(classobj_has_class_repl_key_constraint())
3. 외래키 포함 및 복제 테이블 참조 여부(check_ha_repl_fk_ref_all_replicated())

이전 코드를 포함해 이 세가지 validation 을 수행하는 코드를 리팩토링했고, 이 과정에서 함수check_ha_repl_constraint() 가 추가되었습니다.


ALTER TABLE에서는 fk 수정 시 호출되는 do_alter_one_clause_with_template() 함수에서 validation 을 수행하며, 1, 3번 제약사항에 대해 확인하게됩니다

### Remarks

N/A

youngjun9072 committed:
add replication option

youngjun9072 committed:
lint

youngjun9072 committed:
fix typo

youngjun9072 committed:
Add variable intialization

youngjun9072 committed:
add "alter table ... replication" option

youngjun9072 committed:
lint

youngjun9072 committed:
add "TODO" comment

youngjun9072 committed:
add replication opt handling and select catalog

youngjun9072 committed:
Code changes

- Changed the location where replication option is handled
- Simplified condition logic to check only default or "on"
- Added explanatory comments

youngjun9072 committed:
lint

youngjun9072 committed:
replication 저장 및 조회 구현

youngjun9072 committed:
add alter table handling

youngjun9072 committed:
리뷰 반영
 - 매크로 이름 변경
 - 불필요 공백 삭제

youngjun9072 committed:
에러 핸들링 추가

youngjun9072 committed:
add tab

youngjun9072 committed:
add comment

youngjun9072 committed:
Flag and condition changes for backward compatibility with older backups

- replace SM_CLASSFLAG_REPLICATION with SM_CLASSFLAG_NO_REPLICATION.
- change handling condition from IS_REPLICATION_ON to !IS_REPLICATION.

youngjun9072 committed:
Add macro for alter condition branch

youngjun9072 committed:
Code modification

- Rename flag from SM_CLASSFLAG_REPLICATION to SM_CLASSFLAG_NO_REPLICATION
- Update return values to reflect the new flag naming
- Add error handling when attempting to alter replication tables in HA mode

youngjun9072 committed:
Add error message

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into replication-option-handling

youngjun9072 committed:
remove unused error log

youngjun9072 committed:
Macro refactoring and lint

youngjun9072 committed:
Refactoring

youngjun9072 committed:
lint

youngjun9072 committed:
Apply review feedback
- fix indentation
- rename SM_CLASSFLAG_NO_REPLICATION to SM_CLASSFLAG_REPLICATION_OFF and update comment
- check replication status at function entry to avoid unnecessary operations
- adjust error message and parameter from 1 to 0
- add error handling

youngjun9072 committed:
replication pk only

youngjun9072 committed:
Implementation of unique key insert, delete, and update

youngjun9072 committed:
fix(pgbuf): prevent assertion in pgbuf_is_valid_page during UPDATE/DELETE

youngjun9072 committed:
lint

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into replication_flag_only

youngjun9072 committed:
refectoring

youngjun9072 committed:
lint

youngjun9072 committed:
Fix missing user spec and other updates

- Disallow creation of replication tables without replication key in HA mode
- Prevent REPLICATION ON <-> OFF switching via ALTER in HA mode
  (expected complexity increase with foreign key implementation)
- Rename related macros

youngjun9072 committed:
Review applied

youngjun9072 committed:
Support adding FK constraints on table creation

youngjun9072 committed:
Adding constraints when using ALTER TABLE to add a foreign key (FK) constraint to a column in HA mode.

youngjun9072 committed:
Fix: incorrect replication key detection logic

- Previously, the logic checked constraints and attributes separately.
- This caused a table to be recognized as having a replication key
  even when "UNIQUE" and "NOT NULL" were defined on different columns.
- Updated the logic to require both conditions on the same set of columns.

youngjun9072 committed:
Merge branch 'fix_ha_constraint' into fk_constraint

youngjun9072 committed:
add error handling

youngjun9072 committed:
Merge branch 'feature/add-replication-option' into fk_constraint

youngjun9072 committed:
indent

H2SU commented on src/query/execute_schema.c:
```diff
@@ -8854,6 +8862,36 @@ execute_create_select_query (PARSER_CONTEXT * parser, const char *const class_na
+
+  assert (class_obj != NULL);
+
+  constraints = db_get_constraints (class_obj);
```
따로 constraints를 사용하는 곳이 없는 것 같은데 constraints, tmp_c 둘 중 하나만 사용해도 되지 않을까요?

youngjun9072 replied:
@H2SU 반영했습니다. 더불어 불필요한 변수를 함께 제거하고 코드를 간략화했습니다.

hgryoo commented on src/object/schema_template.c:
```diff
@@ -1664,7 +1664,6 @@ smt_check_foreign_key (SM_TEMPLATE * template_, const char *constraint_name, SM_
   else
     {
       ref_clsop = sm_find_class (fk_info->ref_class);
-
```
관련 없는 수정은 diff로 잡히지 않도록 수정 부탁드립니다.

youngjun9072 replied:
@hgryoo 반영했습니다.

hgryoo commented on src/query/execute_schema.c:
```diff
@@ -8854,6 +8862,36 @@ execute_create_select_query (PARSER_CONTEXT * parser, const char *const class_na
+If the time required to patch the SM is small, the same patch-based approach can be applied consistently. 
+*/
+bool
+ha_fk_replication_violation (DB_OBJECT * class_obj, bool repl)
```
nit: 함수 이름에서 복제 옵션에 대한 확인을 한다는 맥락이 있으면 좋을 것 같습니다.

```suggestion
ha_check_fk_replication_violation (DB_OBJECT * class_obj, bool repl)
```

youngjun9072 replied:
@hgryoo  반영했습니다.

youngjun9072 committed:
Apply code review feedback

- Renamed function ha_fk_replication_violation() to ha_check_fk_replication_violation()
- Removed unnecessary variables and simplified logic in ha_check_fk_replication_violation()

hornetmj commented on src/base/error_code.h:
```diff
@@ -1752,9 +1752,9 @@
 
 #define ER_HA_REQUIRES_REPLICATION_KEY		-1368
 #define ER_REPLICATION_OPTION_CHANGE_IN_HA_MODE		-1369
+#define ER_HA_FK_CONSTRAINT_VIOLATION -1370
```
에러 번호 정렬해 주세요.

youngjun9072 replied:
@hornetmj 반영했습니다. 이 파일에서는 정렬이 탭과 스페이스바를 혼용하고있어 일단 스페이스바로 정렬했고, 별도의 이슈를 따는게 번거로울것같아 이 에픽에서 작성된 다른 에러코드도 함께 정렬했습니다.

hornetmj replied:
@youngjun9072 넵. 수정하신 부분은 space로 정렬하는게 맞습니다. 저희가 set tabstop=8 규칙이고, tab과 space가 혼용되고 있습니다. 기본 .h 파일의 경우 .github/workflows/codestyle.sh에서 수행하는 'indent -l120 -lc120 [수정 파일]' 를 수행해 주시면 됩니다. (구두로 추가 설명드리겠습니다.)

hornetmj commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1463,8 +1463,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 
 1368 In HA mode, a replication key candidate (PRIMARY KEY or NOT NULL UNIQUE constraint) is required when creating a table. If no replication key candidate is defined, set the REPLICATION option to OFF.
 1369 Changing the replication option of a class is not allowed in HA mode.
+1370 In HA mode, the replication option of a table with a foreign key must match that of its referenced table.
```
FK 테이블을 주어로 빼는 것이 에러 메시지 의미 전달이 더 명확할 것 같아 제안드립니다.
```suggestion
1370 In HA mode, a table with a foreign key must have the same replication option as its referenced table.
```
추가적으로, 이 제약들에 대한 메시지에 공통의 헤딩을 붙이는 것도 에러 메시지 의미 전달이 더 명확해 질 것으로 기대됩니다. 적용 검토 요청드립니다.
```suggestion
1370 HA replication constraint violation: A table with a foreign key must have the same replication option as its referenced table.
```

youngjun9072 replied:
@hornetmj 반영했습니다. 추가로 한글 에러 메시지도 유사하게 수정했습니다.

hornetmj commented on src/query/execute_schema.h:
```diff
@@ -107,5 +107,5 @@ extern int do_recreate_filter_index_constr (PARSER_CONTEXT * parser, SM_PREDICAT
 			     int *assigns_count, CLIENT_UPDATE_CLASS_INFO ** cls_data, int *cls_count,
 			     DB_VALUE ** values, int *values_cnt, bool has_delete);
-
+extern bool ha_check_fk_replication_violation (DB_OBJECT * class_obj, bool repl);
```
함수가 execute_schema.c 파일에서만 사용되기 때문에 헤더에 선언되기 보다는 static으로 .c 파일에서 선언되는 것으로 충분할 것 같습니다.

youngjun9072 replied:
@hornetmj 반영하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -9259,6 +9295,14 @@ do_create_entity (PARSER_CONTEXT * parser, PT_NODE * node)
 	}
 
+      if (!HA_DISABLED ()
+	  && ha_check_fk_replication_violation (class_obj, IS_CREATE_STMT_SET_REPL_OPTION (tbl_opt_replication)))
```
위 문맥과 합칠 수 있을 것 같습니다. HA 관련된 제약들을 하나의 문맥으로 합쳐주시면 추후 코드 유지보수에 도움이 될 것 같습니다.

```
// 복제 대상 테이블
if (IS_CREATE_STMT_SET_REPL_OPTION (tbl_opt_replication))
{

}
// 복제 대상이 아닌 테이블
else
{

}
```
위 구조가 기본 골격입니다. 검사해야 하는 부분은 다음과 같습니다.
1. HA 환경에서 복제 대상 테이블은 반드시 RK를 갖어야 한다.
2. HA 환경에서 복제 대상 테이블이 FK를 갖는 경우 참조 테이블(PK 테이블) 역시 복제 대상 테이블이여야 한다. 추가적으로, "현재 생성/수정되는 FK 테이블이 복제 대상이 아니라면, 참조 PK 테이블의 복제 유무는 검사하지 않아도 된다."라고 저는 생각됩니다. 이 부분에 대해 다른 의견 있으시면 말씀해 주세요.

제안 방안:
- check_ha_repl_constraints() 함수를 만들고, 이 함수에서 위 검사들을 수행; !HA_DISABLED () 또한 이 함수의 첫번째 검사 조건으로 넣고, ha 모드가 아닌 경우 성공으로 처리;execute_schema.c 수준에서 정의된 함수는 일관된 함수 prefix가 있어보이지 않고, 로지컬한 네이밍을 사용하고 있음을 고려
- check_ha_repl_constraints() 함수에서 위 두가지 조건 검사에 대해 함수로 분리하거나 각 단계에 대해 주석(ex) /* RULE 1: xxx)을 남겨 명확히 구분하는 것을 추천합니다. 함수로 분리한다면, 주석을 함수 설명에 사용하는 것 역시 고려해 주시면 좋겠습니다.
```
static int
check_ha_repl_constraints(...)
{
    if (HA_DISABLED ())
    {
        return NO_ERROR;
    }

    error = check_ha_repl_has_rk (...);
    또는
    /* RULE 1: In HA mode, a replicated table must have a replication key (RK). This ensures that each row can be uniquely identified during replication. */
    
    error = check_ha_repl_fk_ref_match (...);
    또는
    /* RULE 2: In HA mode, if a replicated table has a foreign key (FK), its referenced (PK) table must also be a replicated table. This prevents replication inconsistency across FK relationships. */

}

if (IS_CREATE_STMT_SET_REPL_OPTION (tbl_opt_replication))
{
    error = check_ha_repl_constraints(...);
    if (error != NO_ERROR)
    {
        // error 처리
    }
}
// 복제 대상이 아닌 테이블
else
{

}
```

youngjun9072 replied:
@hornetmj 제안주신 방안으로 작성할 경우 다음 케이스에대해 논리적 오류가 발생합니다
```
CREATE TABLE a_tbl (
    id INT NOT NULL DEFAULT 0 PRIMARY KEY,
    phone VARCHAR(10)
);

CREATE TABLE b_tbl (
    id INT NOT NULL,
    name VARCHAR (10) NOT NULL,
    CONSTRAINT pk_id PRIMARY KEY (id),
    CONSTRAINT fk_id FOREIGN KEY (id) REFERENCES a_tbl (id)
    ON DELETE CASCADE ON UPDATE RESTRICT
)REPLICATION OFF;
```
스펙상 복제 두 테이블에 복제 옵션이 일치해야하는데, b_tbl은 IS_CREATE_STMT_SET_REPL_OPTION (tbl_opt_replication) 조건식에서 true에 해당이되지않아 복제 옵션이 일치하지 않는 외래 테이블을 참조하게됩니다.

해결방안으로는 '복제 대상이 아닌 테이블'에 대한 코드에도 HA 확인, fk확인 코드가 들어가야하는데 check_ha_repl_constraints 를 다시한번 명시해야해서 이 코드를 따로 빼는게 의미가 있나 싶습니다.

youngjun9072 replied:
@hornetmj 반영하였습니다.

youngjun9072 committed:
Applied review feedback

- Refactored routines for validating replication options and HA-related constraints when creating tables
- Aligned and cleaned up error code definitions
- Updated and clarified related error messages

hornetmj commented on msg/en_US.utf8/cubrid.msg:
```diff
@@ -1463,8 +1463,9 @@ Check the path of the key file (_keys) and make sure that it includes proper key
 
 1368 In HA mode, a replication key candidate (PRIMARY KEY or NOT NULL UNIQUE constraint) is required when creating a table. If no replication key candidate is defined, set the REPLICATION option to OFF.
 1369 Changing the replication option of a class is not allowed in HA mode.
+1370 HA replication constraint violation: A table with a foreign key must have the same replication option as its referenced table.
```
기존 추가된 '복제 제약' 관점의 에러 메시지들 역시 같은 형식으로 정리되면 더 명확한 의미 전달이 될 것 같습니다.

hornetmj replied:
FK 관련 스펙 변경에 따라 에러 메시지도 보다 적합하게 변경을 검토해 봐야 할 것 같습니다. (복제 대상 FK 테이블만 검사되도록 변경)

hornetmj commented on src/base/error_code.h:
```diff
@@ -1750,11 +1750,11 @@
-#define ER_REPLICATION_OPTION_CHANGE_IN_HA_MODE		-1369
-
-#define ER_LAST_ERROR				    -1370
+#define ER_HA_REQUIRES_REPLICATION_KEY              -1368
+#define ER_REPLICATION_OPTION_CHANGE_IN_HA_MODE     -1369
+#define ER_HA_FK_CONSTRAINT_VIOLATION               -1370
```
'복제 제약' 관련 에러 코드가 동일한 형식을 갖는게 어떨까요?
ex) ER_HA_REPL_xxx 등

hornetmj reviewed (approved):
@youngjun9072 구두로 전달드렸던 수정 요청사항이라 커멘트 남기고 승인합니다. 적용하신 후 병합해 주세요.
